### PR TITLE
Pattern matching on cursors

### PR DESCRIPTION
**Pull request by skeqiqevian**
We currently allow users to pattern match via `proc.find`. We will extend that notion to `cursor.find`, enabling pattern matching for cursors.

Also did some miscellaneous code cleanup at the same time. There's a circular import between `API.py` and `API_cursors.py`. I couldn't think of a good way to break the dependency other than merging the two files, anyone have better ideas?

### CODE REVIEW

**Comment by yamaguchi1024** [comment]
Nice. Is this ready for review?

**Comment by skeqiqevian** [comment]
@yamaguchi1024 It should be functional. There are some stylistic choices that are up for debate with regards to code organization.

File: src/exo/API_cursors.py [hunk 1]
```diff
@@ -1005,6 +1011,38 @@ def lift_cursor(impl, proc):
...
+    default_match_no = None if many else 0
+    raw_cursors = match_pattern(
+        scope, pattern, call_depth=1, default_match_no=default_match_no
+    )
+    assert isinstance(raw_cursors, list)
+    cursors = []
+    for c in raw_cursors:
+        c = lift_cursor(c, proc)
+        if isinstance(c, (BlockCursor, ExprListCursor)) and len(c) == 1:
+            c = c[0]
```
Comment (yamaguchi1024) [inline src/exo/API_cursors.py:1037]: This line looks a bit suspicious, isn't this saying that you'll return the first statement of the block if `.find` matches with multiple statements? Can you add a test for querying and scoping with block cursors?

Comment (skeqiqevian) [reply src/exo/API_cursors.py:1037]: That's not what this is doing. Each element of `raw_cursor` is a match. This is more of an artifact of how `pattern_match` interacts with the internal `BlockCursor`. When we match a statement `s` in a body or some expr in a list of accesses (e.g. `i` in `x[i, j, i]`), `match_pattern` returns a list (e.g. `[s]` or `[i]`). This code is to essentially unwrap that before returning it to the user.

Comment (yamaguchi1024) [reply src/exo/API_cursors.py:1037]: I see that makes sense

**Review by yamaguchi1024 — APPROVED** [review]

**Pull request merged**